app/controllers/home_controller.rb

```diff
@@ -1,4 +1,36 @@
 class HomeController < ApplicationController
   def index
+    session[:agency] = current_agency
+    render_agency
+  end
+
+  def success
+    render_agency
+  end
+
+  def failure
+    render_agency
```

Comment:
Was there a particular reason for calling the same method in both success and failure scenarios and putting the logic in the `try_render_template` method? Would it make more sense to treat each event separately? In case of success, render the agency template, and in case of failure, display a specific failure page with an appropriate HTTP status code.